hngprojects/hng_boilerplate_nestjs
# feat:  retrieve users' unread notifications

## What does the PR do?
This PR addresses issue of users to retrieve only their unread notifications for authenticated users from the database.  The endpoint returns errors if:

- [x] User is not logged in.
- [x] User does not exist.
- [x] Validation errors on the user input is done


## How should this be manually tested?
1. Login to a user account, and get the access token

2. You can send a GET request to the following endpoint:
GET /api/v1/notifications.  

## Query Parameters

- is_read=false:

> ```json
> Authorization: Bearer <access_token>
> ```

### Success Response
on a successful execution, you will receive the following response structure:
```
{
  "status": "success",
  "message": "Unread notifications retrieved successfully",
  "status_code": 200,
  "data": {
    "total_notification_count": 5,
    "total_unread_notification_count": 3,
    "notifications": [
      {
        "notification_id": "uuid1",
        "is_read": false,
        "message": "You have a new message",
        "created_at": "2024-07-25T12:00:00Z"
      },
      {
        "notification_id": "uuid2",
        "is_read": false,
        "message": "Your profile has been updated",
        "created_at": "2024-07-24T15:30:00Z"
      }
    ]
  }
}
```

### "Invalid Input" Response
If the request is invalid, you will receive the following response:
```
{
  "status": "error",
  "message": "Invalid request.",
  "status_code": 400
}

```
### Unauthorized Response
```
{
  "status": "error",
  "message": "Unauthorized request",
  "status_code": 401
}
```

### "Internal Server Error" Response

```
{
  "status": "error",
  "message": "Server error",
  "status_code": 500
}

```
###  Manual Testing Instructions

1. Login User:

   - Make a POST request to /api/v1/auth/login with a valid user's email and password.

   - Save the access token from the response.

2. Retrieve Unread Notifications:

    - Make a GET request to /api/v1/notifications?is_read=false with the following:

          Header: Authorization: Bearer <access_token>

3. Verify Response:

      - Ensure the response includes the total count of notifications, total count of unread notifications, and               a list of unread notifications with the correct details.



## Task
- [x]  Implemented GET /api/v1/notifications?is_read=false endpoint in notifications.controller.ts
- [x] Added method to fetch notifications in the NotificationsService.
- [x] Tested endpoint functionality to ensure it handles various scenarios correctly
- [x] Added unit tests for the NotificationsService.
- [x] Updated API documentation to reflect the new endpoint

Reference Issue: [Link to Issue](https://linear.app/team-bingo/issue/BAC2-74/retrieve-users-unread-notifications)
